src/diagram/components/quantity-node.tsx

```diff
@@ -5,27 +5,34 @@ import { Handle, Position } from "react-flow-renderer/nocss";
 import TextareaAutosize from "react-textarea-autosize";
 import classNames from "classnames";
 import { ColorEditor } from "./color-editor";
-import { NumberInput } from "./ui/number-input";
 import { DQNodeType } from "../models/dq-node";
 import { DQRootType } from "../models/dq-root";
-import { kMaxNameCharacters, kMaxNotesCharacters, processName, isValidNumber } from "../utils/validate";
+import { kMaxNameCharacters, kMaxNotesCharacters, processName } from "../utils/validate";
 import { IconColorMenu } from "./icon-color-menu";
 import { IconExpand } from "./icon-expand";
 import { IconWarning } from "./icon-warning";
 
 import "./quantity-node.scss";
+import { ExpandableInput } from "./ui/expandable-input";
 
 interface IProps {
   data: {node: DQNodeType, dqRoot: DQRootType};
   isConnectable: boolean;
 }
 
 const _QuantityNode: React.FC<IProps> = ({ data, isConnectable }) => {
+  const variable = data.node.variable;
+  const kDefaultExpandLength = 10;
+  const kExpressionExpandLength = 18;
+
   const [showColorEditor, setShowColorEditor] = useState(false);
   const [showDescription, setShowDescription] = useState(false);
-  const [showExpression, setShowExpression] = useState(false);
-
-  const variable = data.node.variable;
+  const [showFullExpression, setShowFullExpression] = useState(false);
+  const [showFullValue, setShowFullValue] = useState(false);
+  const [showFullUnits, setShowFullUnits] = useState(false);
+  const [hasLongExpression, setHasLongExpression] = useState(variable.expression && variable.expression.length > 20);
+  const [hasLongValue, setHasLongValue] = useState(variable?.value && variable.value.toString().length > 10);
+  const [hasLongUnits, setHasLongUnits] = useState(variable?.unit && variable.unit.length > 10);
```

Comment:
Can these be turned into local variables rather than state? This would simplify things, plus avoid duplicating the code in two places (see lines 70 and 76).